flink-core/src/main/java/org/apache/flink/configuration/HistoryServerOptions.java

```diff
@@ -99,6 +99,13 @@
 			.withDescription("Enable HTTPs access to the HistoryServer web frontend. This is applicable only when the" +
 				" global SSL flag security.ssl.enabled is set to true.");
 
+	public static final ConfigOption<Integer> HISTORY_SERVER_RETAINED_JOBS =
+		key("historyserver.archive.retained-jobs")
+			.defaultValue(-1)
+			.withDescription(String.format("The maximum number of jobs to retain in each archive directory" +
+				" defined by `%s`. If set to `-1`(default) or negative, there is no limit to the number of archives" +
+				" to keep in the directories and the cache. If set to `0`, no archives are kept.", HISTORY_SERVER_ARCHIVE_DIRS.key()));
```

Comment:
This sounds like 0 should not be an accepted value; there doesn't seem to be a use-case for it.

Reply:
Coming to think about it, it seems to me that a user might want to see only the running jobs in which case they might set `historyserver.archive.retained-jobs = 0` to discard all archives. Although I'm a Flink newbie and I might easily be missing something here.
What do you think @zentol?

Reply:
Hmm, wouldn't `historyserver.archive.retained-jobs = 0` mean that the history server does not show any jobs? I mean it does not make a lot of sense for the user to configure this value tbh. In order to avoid the user misconfiguring this value one could fail if one sees that the user configured it with `0`.

Reply:
Yeah, you are right, I indeed missed that. I might come back and separate the handling of running jobs vs finished jobs under a different PR.

@zentol  @tillrohrmann, could you please help me understand how would you like to see a `0` value handled? I'm a bit confused about what do you mean by failing and not accepting values.
In my opinion, it'd be the easiest and most userfriendly if we'd just fall back to the default disabled mode on every non-positive (<=0) values, i.e. handling `0` and `-1` the same way.

Reply:
I'd throw an `IllegalConfigurationException` if the value is 0 or smaller than -1.
We can keep 0 as a reserved value for the future (e.g., if the HS should show running jobs; or as a way to delete archives continuously without having to redeploy a cluster to change the archiving configuration)
All other negative values should be rejected since they may imply a user-error; say if the file is generated.

Reply:
Failing fast if something is wrongly configured is usually a good idea. One could add an exception message stating that only `-1` or `> 0` values are valid.

Reply:
Thanks for clarifying @zentol and @tillrohrmann, your points make perfect sense. I didn't consider the case of incorrectly generated configs for which I'm all for failing fast.
I also find it cleaner to reserve the `0` value for a later behaviour refinement rather than tolerating the users' error and trying to find out what do they wanted to configure. Besides this makes it easier to test these values. I've updated the PR accordingly.